apps/inheritance_app/inheritance_encrypt_data.c

```diff
@@ -468,23 +597,91 @@ static bool encrypt_data(void) {
+                                       const size_t buffer_len) {
+  size_t total_count = ((buffer_len % ENCRYPTED_CHUNK_SIZE) > 0)
+                           ? (buffer_len / ENCRYPTED_CHUNK_SIZE) + 1
+                           : (buffer_len / ENCRYPTED_CHUNK_SIZE);
+  size_t remaining_size = (size_t)buffer_len;
+  size_t offset = 0;
   inheritance_result_t result =
       init_inheritance_result(INHERITANCE_RESULT_ENCRYPT_TAG);
   result.encrypt.which_response =
-      INHERITANCE_ENCRYPT_DATA_WITH_PIN_RESPONSE_RESULT_TAG;
-  if (!inheritance_get_query(query, INHERITANCE_QUERY_ENCRYPT_TAG) ||
-      !check_which_request(
-          query, INHERITANCE_ENCRYPT_DATA_WITH_PIN_REQUEST_INITIATE_TAG)) {
-    return false;
+      INHERITANCE_ENCRYPT_DATA_WITH_PIN_RESPONSE_ENCRYPTED_DATA_TAG;
+  result.encrypt.encrypted_data.chunk_payload.chunk_index = 0;
+  result.encrypt.encrypted_data.chunk_payload.total_chunks = total_count;
+
+  for (int index = 0; index < total_count; index++) {
+    if (!inheritance_get_query(query, INHERITANCE_QUERY_ENCRYPT_TAG) ||
+        !check_which_request(
+            query,
+            INHERITANCE_ENCRYPT_DATA_WITH_PIN_REQUEST_ENCRYPTED_DATA_REQUEST_TAG)) {
+      return false;
+    }
+    // chunk_payload validation checks
+    if (query->encrypt.encrypted_data_request.has_chunk_ack == false ||
+        query->encrypt.encrypted_data_request.chunk_ack.chunk_index != index) {
+      return false;
+    }
+    size_t chunk_size = (remaining_size > ENCRYPTED_CHUNK_SIZE)
+                            ? ENCRYPTED_CHUNK_SIZE
+                            : remaining_size;
+    remaining_size -= chunk_size;
+    result.encrypt.encrypted_data.chunk_payload.remaining_size = remaining_size;
+    result.encrypt.encrypted_data.has_chunk_payload = true;
+    memcpy(result.encrypt.encrypted_data.chunk_payload.chunk.bytes,
+           buffer + offset,
+           chunk_size);
+    result.encrypt.encrypted_data.chunk_payload.chunk.size = chunk_size;
+    inheritance_send_result(&result);
+    offset += chunk_size;
+    result.encrypt.encrypted_data.chunk_payload.chunk_index++;
+    if (remaining_size == 0) {
+      break;
+    }
   }
+  return true;
+}
 
-  result.encrypt.result.encrypted_data.size = encryption_context->packet_size;
-  memcpy(&result.encrypt.result.encrypted_data.bytes,
-         encryption_context->packet,
-         encryption_context->packet_size);
-
-  inheritance_send_result(&result);
+static bool send_encrypted_data(inheritance_query_t *query) {
+  uint8_t
+      buffer[INHERITANCE_ENCRYPT_DATA_WITH_PIN_ENCRYPTED_DATA_STRUCTURE_SIZE] =
+          {0};
+  size_t bytes_encoded = 0;
+  if (!get_pb_encoded_buffer(&encryption_context->payload,
+                             buffer,
+                             sizeof(buffer),
+                             &bytes_encoded)) {
+    return false;
+  }
+  if (!inheritance_send_in_chunks(query, buffer, bytes_encoded)) {
+    return false;
```

Comment:
We can merge the if conditions and add a todo to throw an error incase false is returned